operations_on_datastructures/inorder_successor_of_bst.cpp

```diff
@@ -0,0 +1,407 @@
+ * */
+Node *makeNode(int64_t data) {
+    Node *node = new Node();
+    node->data = data;      ///< setting data for node
+    node->left = nullptr;   ///< setting left child as null
+    node->right = nullptr;  ///< setting right child as null
+    return node;
+}
+
+/**
+ * @brief Inserts the given data in BST while maintaining the properties of BST.
+ * @param root Pointer to the root node of the BST
+ * @param data Data to be inserted.
+ * @returns Node* Pointer to the root node.
+ * */
+Node *Insert(Node *root, int64_t data) {
+    if (root == nullptr) {
+        root = makeNode(data);
+    } else if (data <= root->data) {
+        root->left = Insert(root->left, data);
+    } else {
+        root->right = Insert(root->right, data);
+    }
+    return root;
+}
+
+/**
+ * @brief Searches the given data in BST and returns the pointer to the node
+ * containing that data.
+ * @param root Pointer to the root node of the BST
+ * @param data Data to be Searched.
+ * @returns Node* pointer to the found node
+ * */
+Node *getNode(Node *root, int64_t data) {
+    if (root == nullptr) {
+        return nullptr;
+    } else if (root->data == data) {
+        return root;
+    } else if (data > root->data) {
+        /// recursive call
+        return getNode(root->right, data);
+    } else {
+        /// recursive call
+        return getNode(root->left, data);
+    }
+}
+
+/**
+ * @brief Finds and return the minimum node in BST.
+ * @param root A pointer to root node.
+ * @returns Node* Pointer to the found node
+ * */
+Node *findMinNode(Node *root) {
+    if (root == nullptr) {
+        return root;
+    }
+    while (root->left != nullptr) {
+        root = root->left;
+    }
+    return root;
+}
+
+/**
+ * @brief Prints the BST in inorder traversal using recursion.
+ * @param root A pointer to the root node of the BST.
+ * @returns void
+ * */
+void printInorder(Node *root) {
+    if (root == nullptr) {
+        return;
+    }
+
+    printInorder(root->left);  /// recursive call to left subtree
+    std::cout << root->data << " ";
+    printInorder(root->right);  /// recursive call to right subtree
+}
+
+/**
+ * @brief This function is used in test cases to quickly create BST containing
+ * large data instead of hard coding it in code. For a given root, this will add
+ * all the nodes containing data passes in data vector.
+ * @param root Pointer to the root node.
+ * @param data A vector containing integer values which are suppose to be
+ * inserted as nodes in BST.
+ * @returns Node pointer to the root node.
+ * */
+Node *makeBST(Node *root, const std::vector<int64_t> &data) {
+    for (int64_t values : data) {
+        root = Insert(root, values);
+    }
+    return root;
+}
+
+/**
+ * @brief Search from the root node as we need to walk the tree starting from
+ * the root node to the given node, by doing so, we are visiting every ancestor
+ * of the given node. In order successor would be the deepest node in this path
```

Comment:
Hmm, I guess some part of this *brief* description should be moved to a *detailed* description.

Reply:
yea, that's too much for a brief intro 🤓, I'll do it!